nbextensions/bulkImportCell/bulkImportCell.js

```diff
@@ -761,6 +769,52 @@ define([
             }
         }
 
+        /**
+         * Starts the cell by executing the generated code in its input area.
+         * Then changes the global cell state to "launching".
+         */
+        function doRunCellAction() {
+            runStatusListener = cellBus.on('run-status', handleRunStatus);
+            busEventManager.add(runStatusListener);
+            cell.execute();
+            updateState('launching');
+        }
+
+        /**
+         * Globally cancels all running jobs and the run status by the following steps.
+         * 1. If we've clicked run, but we don't have any jobs yet, just remove the
+         *    listener and ignore the jobs.
+         *    //TODO: make this wait for jobs to start first, then cancel them all in turn?
```

Comment:
I'm a little unsure what to do in the case where the user starts the job, then immediately clicks cancel before the job actually starts. Right now, it just ignores any run-status commands from here on, effectively letting them all run without watching them, which might be a problem if the user fires off 100s of imports then cancels.

I think we need to do something extra clever here, but I'm not sure the best way, yet.

Reply:
This will also be an issue with retrying jobs and/or if there is a connectivity failure that means vital job messages are missed. This could be managed in the backend, having the backend tracking the actions that the user has taken and following up on any inconsistencies - expected responses from ee2 that haven't arrived, if the job has been started already, etc.